### PR TITLE
esptool: 3.3.1 -> 4.4

### PR DESCRIPTION
**Pull request by dotlambda**
###### Description of changes
https://github.com/espressif/esptool/releases/tag/v4.0

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Tested with esphome, work for me, but maybe we can wait for https://github.com/esphome/esphome/pull/3481 to rely on upstream to rule out possible problems?

**Comment by yorickvP** [comment]
@SnoozeThis https://github.com/esphome/esphome/pull/3504 || wait for 1 week

**Comment by SnoozeThis** [comment]
(https://snoozeth.is/PDP083q3770) I will either wait until https://github.com/esphome/esphome/pull/3504 is merged or until Fri, 10 Jun 2022 16:14:37 UTC and then add a comment.

**Comment by mweinelt** [comment]
I think this will be a blocker, since the API has changed somewhat and the esphome CI shows the entrypoint isn't the same anymore.

```
### File esphome/__main__.py
Error: esphome/__main__.py:245: [E1101(no-member), upload_using_esptool.run_esptool] Module 'esptool' has no '_main' member

**Comment by SnoozeThis** [comment]
Resolved.

**Comment by yorickvP** [comment]
Should we split to esptool3 and esptool4?

**Comment by mweinelt** [comment]
Yeah, I think splitting could be a solution for the time being.

**Comment by mweinelt** [comment]
I think the split should happen soon, so we'll have both in 22.11.

File: pkgs/tools/misc/esptool/3.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    ${python3.interpreter} test/test_modules.py
+
+    runHook postCheck
+  '';
+
+  meta = with lib; {
+    description = "ESP8266 and ESP32 serial bootloader utility";
+    homepage = "https://github.com/espressif/esptool";
+    license = licenses.gpl2Plus;
+    maintainers = with maintainers; [ ];
```
Comment (dotlambda) [inline pkgs/tools/misc/esptool/3.nix:60]: @mweinelt? @yorickvP?

Comment (mweinelt) [reply pkgs/tools/misc/esptool/3.nix:60]: I can take it.

Comment (mweinelt) [reply pkgs/tools/misc/esptool/3.nix]: hexa :)

**Pull request merged**